### PR TITLE
orm insert or update

### PR DESCRIPTION
**Pull request by fudali113**

### CODE REVIEW

**Comment by fudali113** [comment]
@astaxie
finally passed the test 😃

File: orm/orm.go [hunk 1]
```diff
@@ -209,6 +209,19 @@ func (o *orm) InsertMulti(bulk int, mds interface{}) (int64, error) {
 	return cnt, nil
 }
 
+//insert or update data to database
```
Comment (astaxie) [inline orm/orm.go]: comments should be start with `// InsertOrUpdate`

File: orm/db.go [hunk 2]
```diff
@@ -488,6 +488,114 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
 	return id, err
 }
 
+//insert or update a row
+//If your primary key or unique column conflict will update
+//if no will insert
```
Comment (astaxie) [inline orm/db.go]: comments should be start with `// InsertOrUpdate `

File: orm/types.go [hunk 3]
```diff
@@ -53,6 +53,11 @@ type Ormer interface {
 	//  id, err = Ormer.Insert(user)
 	//  user must a pointer and Insert will set user's pk field
 	Insert(interface{}) (int64, error)
+	//mysql:InsertOrUpdate(model) or InsertOrUpdate(model,"colu=colu+value")
```
Comment (astaxie) [inline orm/types.go]: add a space after the `//`

**Comment by fudali113** [comment]
@astaxie ok

File: orm/orm.go [hunk 4]
```diff
@@ -209,6 +209,19 @@ func (o *orm) InsertMulti(bulk int, mds interface{}) (int64, error) {
 	return cnt, nil
 }
 
+// InsertOrUpdate data to database
+func (o *orm) InsertOrUpdate(md interface{}, colConflitAndArgs ...string) (int64, error) {
+	mi, ind := o.getMiInd(md, true)
+	id, err := o.alias.DbBaser.InsertOrUpdate(o.db, mi, ind, o.alias.TZ, o.alias.DriverName, colConflitAndArgs...)
```
Comment (astaxie) [inline orm/orm.go]: o.alias.DriverName change to o.alias.DriverType. As we already has define https://github.com/astaxie/beego/blob/master/orm/db_alias.go#L29-L36. So you can remove the hardcode mysql and postgres. As sometime the DriverName is not equal to mysql and postgres.

**Comment by fudali113** [comment]
@astaxie ok

File: orm/db.go [hunk 5]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
 	return id, err
 }
 
+// InsertOrUpdate a row
+// If your primary key or unique column conflict will update
+// If no will insert
+func (d *dbBase) InsertOrUpdate(q dbQuerier, mi *modelInfo, ind reflect.Value, a *alias, args ...string) (int64, error) {
+	iouStr := ""
+	mysql := DRMySQL
+	postgres := DRPostgres
```
Comment (astaxie) [inline orm/db.go]: why you still need these two params?

File: orm/db.go [hunk 6]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
...
+	driver := a.Driver
```
Comment (astaxie) [inline orm/db.go]: why not just use `a.Driver == DRMySQL`?

File: orm/db.go [hunk 7]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
...
+	argsMap := map[string]string{}
+	args0 := ""
+	if driver == mysql {
+		iouStr = "ON DUPLICATE KEY UPDATE"
+	} else if driver == postgres {
+		if len(args) == 0 || (len(strings.Split(args0, "=")) != 1) {
+			return 0, fmt.Errorf("`%s` use insert or update must have a conflict column arg in first", a.DriverName)
```
Comment (astaxie) [inline orm/db.go]: update the error message `use InsertOrUpdate must have a confilict column`

File: orm/db.go [hunk 8]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
...
+	iouStr := ""
+	mysql := DRMySQL
+	postgres := DRPostgres
+	driver := a.Driver
+	argsMap := map[string]string{}
+	args0 := ""
+	if driver == mysql {
+		iouStr = "ON DUPLICATE KEY UPDATE"
+	} else if driver == postgres {
+		if len(args) == 0 || (len(strings.Split(args0, "=")) != 1) {
```
Comment (astaxie) [inline orm/db.go]: why add this judgement `(len(strings.Split(args0, "=")) != 1`

File: orm/db.go [hunk 9]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
...
+			return 0, fmt.Errorf("`%s` use insert or update must have a conflict column arg in first", a.DriverName)
+		} else {
+			args0 = strings.ToLower(args[0])
+			iouStr = fmt.Sprintf("ON CONFLICT (%s) DO UPDATE SET", args0)
```
Comment (astaxie) [inline orm/db.go]: I  think you can just use `iouStr = fmt.Sprintf("ON CONFLICT (%s) DO UPDATE SET", strings.ToLower(args[0]))`

File: orm/db.go [hunk 10]
```diff
@@ -488,6 +488,115 @@ func (d *dbBase) InsertValue(q dbQuerier, mi *modelInfo, isMulti bool, names []s
...
+		}
+	} else {
+		return 0, fmt.Errorf("`%s` nonsupport insert or update in beego", a.DriverName)
```
Comment (astaxie) [inline orm/db.go]: `insert or update` change to the `InsertOrUpdate`

File: orm/db.go (hunk 6 above)
Comment (fudali113) [inline orm/db.go]: 是的，我今天改的时候没注意到他们已经没有意义了

File: orm/db.go (hunk 8 above)
Comment (fudali113) [inline orm/db.go]: 为了用户在传了“colu=colu+1”时能有更友好的提示

File: orm/db.go (hunk 9 above)
Comment (fudali113) [inline orm/db.go]: `
if vtl == args0 {
	conflitValue = values[i]
}
`这里还有用，因为names里的值应该全是小写的

File: orm/db.go (hunk 7 above)
Comment (fudali113) [inline orm/db.go]: @astaxie 
ok

File: orm/db.go (hunk 10 above)
Comment (fudali113) [inline orm/db.go]: ok

**Comment by fudali113** [comment]
@astaxie 
可以看看现在还有什么地方需要修改么？

**Comment by astaxie** [comment]
LGTM

**Pull request merged**